packages/devtools_app/lib/src/charts/chart_controller.dart

```diff
@@ -0,0 +1,449 @@
+
+/// Zoom level of chart
+enum displayDuration { live, fiveMinutes, fifteenMinutes, all }
+
+/// Given a durationLabel return the displayDuration.
+displayDuration durationToDisplayDuration(
+  displayDuration currentZoomDuration,
+  Duration duration,
+) {
+  if (currentZoomDuration == displayDuration.live) {
+    if (duration.inMinutes == 0 && duration.inSeconds >= 10) {
+      return displayDuration.live;
+    }
+  } else if (currentZoomDuration == displayDuration.fifteenMinutes) {
+    if (duration.inMinutes >= 5) {
+      return displayDuration.fifteenMinutes;
+    }
+  } else if (currentZoomDuration == displayDuration.fiveMinutes) {
+    if (duration.inMinutes == 3) {
+      return displayDuration.fiveMinutes;
+    }
+  }
+
+  return null;
+}
+
+String prettyTimestamp(int timestamp) {
+  final timestampDT = DateTime.fromMillisecondsSinceEpoch(timestamp);
+  return intl.DateFormat.Hms().format(timestampDT); // HH:mm:ss
+}
+
+/// Indexes into xAxisLabeledTimestamps.
+const leftLabelIndex = 0;
+const centerLabelIndex = 1;
+const rightLabelIndex = 2;
+
+/*
+                                          xPaddingRight
+                                          |
+      leftPadding          topPadding     |  rightPadding
+      |                    |              |  |
+    ==|====================|==============|==|=
+    " |                    V              |  |
+    " V ----------------------------------V  V
+    "   |
+    "   |
+    "   |
+    "   | ------ canvasChartWidth --------
+    "   |
+    "   |
+    "   ---------------------------------- <----- yCanvasChart
+    "   '  '  '  '  '  '  '  '  '  '  '  '    {-------- bottomPadding
+    "   ^                ^                     \
+    ====|================|======================
+        |                |
+        xCanvasChart     X-axis ticks
+
+*/
+class ChartController extends DisposableController
+    with AutoDisposeControllerMixin {
+  ChartController({
+    this.displayXLabels = true,
+    this.displayYLabels = true,
+  });
+
+  // Total size of area to display the chart (includes title, axis, labels, etc.).
+  Size size;
+
+  // TODO(terry): Compute dynamically based on title text height.
+  final topPadding = 20.0;
+  // TODO(terry): Compute dynamically based on Y-axis lables text width.
+  final leftPadding = 50.0;
+  // Computed minimum right padding.
+  int rightPadding = 25;
+  // TODO(terry): Compute dynamically based on X-axis lables text height.
+  final bottomPadding = 40.0;
+
+  // TODO(terry): Consider computation of each X-axis tick width (scaled).
+  double tickWidth = 10.0;
+
+  int visibleTicks;
+
+  // TODO(terry): For now three labels.  Need better mechanism, some number of labels
+  //              based on x-axis zoom factor and default unit to display for labels e.g.,
+  //              live take the duration (visible) and divide by some nice unit e.g., 10/20
+  //              seconds, 5 minute show 1 minute units, 15 minutes show 5 minute units, etc.
```

Comment:
should also adjust the # of labels based on how wide the window is.

Reply:
maybe also add a comment by the left, center, and right label code to make it clear we know it should be rewritten.

Reply:
Done.